### ISSUE TITLE
Hyperparameter optimisation

### ISSUE THREAD

**Issue by GMCobraz** [caponetto/bayesian-hierarchical-clustering#3]
Dear caponetto and others,

Good day,
i am working on the hyperparameter optimisation.
I have some questions.
May I know, what is the g refers to in prior.py?
scatter_matrix = (data_matrix_cov / g).T

Thank you very much

**Comment by caponetto** [comment]
Hi @GMCobraz, 
The way the prior is being created [here](https://github.com/caponetto/bayesian-hierarchical-clustering/blob/main/bhc/core/prior.py#L60) comes from the original matlab code, where `g scales empirical prior cov S`.

**Comment by GMCobraz** [comment]
Hi @caponetto ,
Thanks for the reply.
![image](https://user-images.githubusercontent.com/63672255/118388985-d5178400-b627-11eb-925e-6f3dea5992cc.png)
Based on my understanding, cov_matrix = g * scatter_matrix, where g is 1/(N-1).
I am not understand why g for bhc and brt are different for the example, since N=18.

May I know why in prior.py, degrees_of_freedom = data.shape[1] + 1?

Is the scale factor can be understand as learning rate in this case?

Thank you very much.

**Comment by GMCobraz** [comment]
Hi @caponetto ,

May I know do you have an idea to implement the hyperparameter optimisation?
I am stuck halfway in implementing the brent method to optimize gamma for rose tree

Thank you very much.

**Comment by caponetto** [comment]
Hi @GMCobraz,

Unfortunately, I don't have details about the behavior of the hyperparameters.
Maybe this is a good subject for exploring as part of your research?
If you do so, please share the results. :)

Also, notice that the example was taken from the original paper.
So I suppose that the hyperparameters are optimized for BHC.
For BRT, on the other hand, they are not optimized.
My goal was to use the same example for both strategies and provide an intuition.

Regarding the hyperparameter optimization itself, I suggest you take a look at Random Search [1].

[1] https://www.jmlr.org/papers/volume13/bergstra12a/bergstra12a.pdf
